### ISSUE TITLE
@UtilityClass should produce a compiler error, when put on a non-static inner class

### ISSUE THREAD

**Issue by tomsit-ionos** [projectlombok/lombok#3778]
## TLDR
When adding `@UtilityClass` **on a non-static inner class** with Java 21, then this causes VM crashes and other non-obvious problems.

If u add the `static` modifier, it works (for me).

See my stripped down reproduction repo @ https://github.com/tomsit-ionos/lombok-utilclass-bug

Also notice that the intellij decompiler is not able to open generated class-file and  i verified this also with an online decompiler.

## Long story
For me it was a nested class **inside** a record. 

It took me a good day to figure out what's going on b/c i ran into this while upgrade a keycloak extension from KC 25 -> 26.
This set me on the wrong mental path and i was guessing there were some wired upgrade problems at play.

To add to the fun: 
The nested class didnt pose a problem for non-testcontainer junit tests using that inner class, ie. these were green! 
But only **with the intellij junit runner** -- which is what I typically use, while working on code, due to the better DX.

What i failed to notice was, that theses normal tests failed when executed by `mvn test`  (e.g. on CI) with this message:
```
...
[ERROR] The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
...

```

_Originally posted by @tomsit-ionos in https://github.com/projectlombok/lombok/issues/3679#issuecomment-2466305132_

**Comment by Rawi01** [comment]
Interesting, I decompiled the generated class file and the output is different between javac 17 and 21. This difference is also present at the bytecode level.

```diff
   ...
   public static final class InnerClass {
      public static final String BAR = "bar";
-     // $FF: synthetic field
-     final OuterClass this$0;

      @Generated
      private InnerClass(OuterClass var1) {
-        this.this$0 = var1;
         throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
      }
   }
```

**Comment by Rawi01** [comment]
TIL: `javac` does no longer add references to unused outer classes for targets >= Java 18. So this does not cause the problem.

**Comment by Rawi01** [comment]
I cannot build a class file that is broken, can you reproduce it on another system? Can you also please share the Java version you are using for your build?

**Comment by JakubWagner1019** [comment]
> TIL: `javac` does no longer add references to unused outer classes for targets >= Java 18. So this does not cause the problem.

https://bugs.openjdk.org/browse/JDK-8271717

**Comment by briacp** [comment]
We stumbled into the same bug yesterday and it left us dumbfounded for a while. Here's a minimal test case to reproduce it:

```java
package com.example;

import lombok.experimental.UtilityClass;

public class UtilityClassInsideInnerClassJdk21 {

    @UtilityClass
    private class UtilityInnerClass {
        public static String callInnerClassMethod() {
            return "Great, it works!";
        }
    }
}

```

Building this class with gradle doesn't yield any errors, but the resulting `.class` files can't be executed nor decompiled.

Changing the `private class` to `private static` solves the issue.

```
❯ java -version
openjdk version "21.0.6" 2025-01-21
OpenJDK Runtime Environment Homebrew (build 21.0.6)
OpenJDK 64-Bit Server VM Homebrew (build 21.0.6, mixed mode, sharing)
```

**Comment by briacp** [comment]
Nevermind, it's fixed in 1.18.38! Thanks for your work 👍

**Comment by rzwitserloot** [comment]
@briacp That was all @Rawi01 's doing. It's a bizarre little corner case.
